### PR TITLE
Feature t1 33 add email existance verification

### PR DESCRIPTION
**Pull request by tarurar**

### CODE REVIEW

File: src/Core/Exceptions/BCryptWorkFactorInconsistencyException.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        public BCryptWorkFactorInconsistencyException(string message, Exception innerException) : base(message, innerException)
+        {
+        }
+
+        protected BCryptWorkFactorInconsistencyException(SerializationInfo info, StreamingContext context) : base(info, context)
+        {
+        }
+
+        public int WorkFactor { get; set; }
```
Comment (gmaf) [inline src/Core/Exceptions/BCryptWorkFactorInconsistencyException.cs]: * Remove setter. 
* Lift field above constructor in code.

File: src/Core/Exceptions/BCryptWorkFactorInvalidException.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Core.Exceptions
+{
+    public class BCryptWorkFactorInvalidException : Exception
+    {
+        public BCryptWorkFactorInvalidException()
+        {
+        }
+
+        public BCryptWorkFactorInvalidException(int workFactor) : base("Work factor is invalid")
```
Comment (gmaf) [inline src/Core/Exceptions/BCryptWorkFactorInvalidException.cs]: Same as for ```BCryptWorkFactorInconsistencyException ```

File: src/Core/Exceptions/EmailHashInvalidException.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Core.Exceptions
+{
+    public class EmailHashInvalidException : Exception
+    {
+        public EmailHashInvalidException()
+        {
+        }
+
+        public EmailHashInvalidException(string email) : base("Hash is invalid")
```
Comment (gmaf) [inline src/Core/Exceptions/EmailHashInvalidException.cs]: Same as for ```BCryptWorkFactorInconsistencyException ```
If you want email, add constructor which could accept message and email. 
But message would be optional in it.

File: src/Core/Extensions/BCryptExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+
+namespace Core.Extensions
+{
+    public static class BCryptExtensions
```
Comment (gmaf) [inline src/Core/Extensions/BCryptExtensions.cs]: Please add ```<summary>```

File: src/Core/Extensions/BCryptExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public static char BCryptTokenSeparator = '$';
```
Comment (gmaf) [inline src/Core/Extensions/BCryptExtensions.cs]: Please add ```<summary>``` for public field.

File: src/Core/Extensions/BCryptExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+        /// <summary>
+        /// Extracts bCrypt work factor from hash string
+        /// </summary>
+        /// <param name="src">The hash string</param>
+        /// <returns>Work factor value</returns>
+        /// <exception cref="ArgumentNullException"></exception>
```
Comment (gmaf) [inline src/Core/Extensions/BCryptExtensions.cs]: Please add condition description for ```ArgumentNullException```, starting with:
> Thrown when ...

File: src/Core/Services/IBCryptService.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+﻿using Core.Exceptions;
+
+namespace Core.Services
+{
+    public interface IBCryptService
```
Comment (gmaf) [inline src/Core/Services/IBCryptService.cs]: Please add ```<summary>``` for interface.

File: src/Core/Services/IEmailValidationService.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading.Tasks;
+
+namespace Core.Services
+{
+    public interface IEmailValidationService
```
Comment (gmaf) [inline src/Core/Services/IEmailValidationService.cs]: Please add ```<summary>``` for interface.

File: src/Core/Services/IEmailValidationService.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+    {
+        Task<bool> IsEmailTakenAsync(string email, string hash);
```
Comment (gmaf) [inline src/Core/Services/IEmailValidationService.cs]: * Please add ```<summary>``` for method.
* Please describe possible exceptions here too.

File: src/Lykke.Service.OAuth.Services/EmailValidationService.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Threading.Tasks;
+using Core.Services;
+using JetBrains.Annotations;
+using Lykke.Service.ClientAccount.Client;
+using Lykke.Service.ClientAccount.Client.Models;
+
+namespace Lykke.Service.OAuth.Services
+{
+    public class EmailValidationService : IEmailValidationService
```
Comment (gmaf) [inline src/Lykke.Service.OAuth.Services/EmailValidationService.cs]: Please add ```/// <inheritdoc />``` for ```EmailValidationService ```.

File: src/Lykke.Service.OAuth.Services/EmailValidationService.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        /// <summary>
+        /// Checks if email already used
+        /// </summary>
+        /// <param name="email"></param>
+        /// <param name="hash"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public async Task<bool> IsEmailTakenAsync(string email, string hash)
```
Comment (gmaf) [inline src/Lykke.Service.OAuth.Services/EmailValidationService.cs]: Please add ```/// <inheritdoc />``` for ```IsEmailTakenAsync```, and remove duplicate xml comments.

File: src/Lykke.Service.OAuth.Services/BCryptService.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+﻿using Core.Exceptions;
+using Core.Extensions;
+using Core.Services;
+
+namespace Lykke.Service.OAuth.Services
+{
+    public class BCryptService : IBCryptService
```
Comment (gmaf) [inline src/Lykke.Service.OAuth.Services/BCryptService.cs]: Please add ```/// <inheritdoc />``` for ```BCryptService```.

File: src/Lykke.Service.OAuth.Services/BCryptService.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        private readonly int _bCryptWorkFactorSettings;
+
+        public BCryptService(int bCryptWorkFactorSettings)
+        {
+            _bCryptWorkFactorSettings = bCryptWorkFactorSettings;
+        }
+
+        public void Verify(string source, string hash)
```
Comment (gmaf) [inline src/Lykke.Service.OAuth.Services/BCryptService.cs]: Please add ```/// <inheritdoc />``` for ```Verify```.

File: src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs [hunk 14]
```diff
@@ -288,6 +298,40 @@ public async Task<VerificationCodeResult> VerifyEmail([FromBody] VerificationCod
...
+        /// Check if email is already registered
+        /// </summary>
+        /// <param name="model"></param>
+        /// <response code="200">Validation result</response>
+        /// <response code="400">Email hash is invalid, BCrypt work factor is invalid</response>
+        [HttpPost]
+        [Route("~/registration/email")]
+        [SwaggerOperation("ValidateEmail")]
+        [ProducesResponseType(typeof(ErrorResponse), (int) HttpStatusCode.BadRequest)]
+        [ProducesResponseType(typeof(EmailValidationResult), (int) HttpStatusCode.OK)]
```
Comment (gmaf) [inline src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs]: Better to move successful response Attribute before unsuccessful ones.
```suggestion
        [ProducesResponseType(typeof(EmailValidationResult), (int) HttpStatusCode.OK)]
        [ProducesResponseType(typeof(ErrorResponse), (int) HttpStatusCode.BadRequest)]
```

File: src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs [hunk 15]
```diff
@@ -288,6 +298,40 @@ public async Task<VerificationCodeResult> VerifyEmail([FromBody] VerificationCod
...
+        [ValidateModel]
+        public async Task<IActionResult> ValidateEmail([FromBody] ValidateEmailModel model)
```
Comment (gmaf) [inline src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs]: Better to rename ```ValidateEmailModel``` to 
* ```ValidateEmailRequest```
* or ```ValidateEmailRequestModel```

File: src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs [hunk 16]
```diff
@@ -288,6 +298,40 @@ public async Task<VerificationCodeResult> VerifyEmail([FromBody] VerificationCod
...
+        {
+            try
+            {
+                bool isEmailTaken = await _emailValidationService.IsEmailTakenAsync(model.Email, model.Hash);
+
+                return Ok(new EmailValidationResult {IsEmailTaken = isEmailTaken});
+            }
+            catch (EmailHashInvalidException e)
+            {
+                _log.Warning("Invalid hash has been provided for email", e, $"email = {e.Email}");
```
Comment (gmaf) [inline src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs]: I think we should not log user emails.
We could just log that ```"Invalid hash has been provided for email"```.
[Discussable]

File: src/Lykke.Service.OAuth/Models/ValidateEmailModel.cs [hunk 17]
```diff
@@ -0,0 +1,23 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Lykke.Service.OAuth.Models
+{
+    /// <summary>
+    /// Email validation request model
+    /// </summary>
+    public class ValidateEmailModel
```
Comment (gmaf) [inline src/Lykke.Service.OAuth/Models/ValidateEmailModel.cs]: Better to rename ```ValidateEmailModel``` to 
* ```ValidateEmailRequest```
* or ```ValidateEmailRequestModel```

File: src/Lykke.Service.OAuth/Startup.cs [hunk 18]
```diff
@@ -166,11 +179,13 @@ public IServiceProvider ConfigureServices(IServiceCollection services)
...
                         $"{DataProtectionContainerName}/cookie-keys/keys.xml");
 
+                services.AddSwaggerGen(opt => opt.DefaultLykkeConfiguration("v1", "Lykke OAuth Server"));
+
                 builder.RegisterModule(new WebModule(settings));
                 builder.RegisterModule(new DbModule(settings));
                 builder.RegisterModule(new BusinessModule(settings));
                 builder.RegisterModule(new ClientServiceModule(settings));
-                builder.RegisterModule(new ServiceModule());
+                builder.RegisterModule(new ServiceModule(settings.CurrentValue.OAuth.Security.BCryptWorkFactor));
```
Comment (gmaf) [inline src/Lykke.Service.OAuth/Startup.cs]: It's better to pass ```settings``` here and then get ```BCryptWorkFactor``` inside module.

Comment (tarurar) [reply src/Lykke.Service.OAuth/Startup.cs]: I'd pass only necessary settings. No need to pass all the complex object.

File: src/Core/Extensions/BCryptExtensions.cs (hunk 6 above)
Comment (gmaf) [reply src/Core/Extensions/BCryptExtensions.cs]: Still needs condition description. 
```
        /// <exception cref="ArgumentNullException">Here</exception>
```

File: src/Core/Services/IBCryptService.cs (hunk 7 above)
Comment (gmaf) [reply src/Core/Services/IBCryptService.cs]: Done

File: src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs (hunk 16 above)
Comment (tarurar) [reply src/Lykke.Service.OAuth/Controllers/AuthenticationController.cs]: Decided not to remove this email from log

**Pull request merged**